File: frontend/src/cacheWarmup/prefetchPlanner.js
const clamp = (value, min, max) => Math.min(max, Math.max(min, value));

const percentile = (values, p) => {
  if (!Array.isArray(values) || values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const index = clamp(Math.floor((sorted.length - 1) * p), 0, sorted.length - 1);
  return sorted[index];
};

export const buildPrefetchPlan = ({
  latencies = [],
  missRate = 0,
  hotKeys = [],
  concurrency = 4,
}) => {
  const p90Latency = percentile(latencies, 0.9);
  const normalizedMissRate = clamp(missRate, 0, 1);
  const budget = Math.round(clamp(12 + normalizedMissRate * 40 - p90Latency / 40, 4, 40));
  const laneCount = clamp(concurrency, 1, 8);

  const selectedKeys = hotKeys.slice(0, budget);
  const lanes = Array.from({ length: laneCount }, () => []);

  selectedKeys.forEach((key, index) => {
    lanes[index % laneCount].push(key);
  });

  return {
    budget,
    laneCount,
    p90Latency,
    lanes,
  };
};

export const estimateWarmupWindowMs = ({ budget = 0, laneCount = 1, p90Latency = 0 }) => {
  const effectiveLanes = clamp(laneCount, 1, 8);
  const batches = Math.ceil(clamp(budget, 0, 200) / effectiveLanes);
  return batches * clamp(Math.round(p90Latency), 10, 500);
};

File: backend/utils/cache_warmup/prefetch_planner.py
from dataclasses import dataclass
from math import ceil
from typing import Iterable, List


def _clamp(value: float, lower: float, upper: float) -> float:
    return max(lower, min(value, upper))


def _percentile(values: List[float], p: float) -> float:
    if not values:
        return 0.0
    sorted_values = sorted(values)
    index = int(_clamp((len(sorted_values) - 1) * p, 0, len(sorted_values) - 1))
    return float(sorted_values[index])


@dataclass(frozen=True)
class PrefetchSignal:
    latencies: List[float]
    miss_rate: float
    hot_keys: List[str]
    concurrency: int = 4


@dataclass(frozen=True)
class PrefetchPlan:
    budget: int
    lane_count: int
    p90_latency: float
    lanes: List[List[str]]


def build_prefetch_plan(signal: PrefetchSignal) -> PrefetchPlan:
    p90_latency = _percentile(signal.latencies, 0.9)
    normalized_miss_rate = _clamp(signal.miss_rate, 0.0, 1.0)
    budget = int(round(_clamp(12 + normalized_miss_rate * 40 - p90_latency / 40, 4, 40)))
    lane_count = int(_clamp(signal.concurrency, 1, 8))

    selected_keys = signal.hot_keys[:budget]
    lanes: List[List[str]] = [[] for _ in range(lane_count)]

    for index, key in enumerate(selected_keys):
        lanes[index % lane_count].append(key)

    return PrefetchPlan(
        budget=budget,
        lane_count=lane_count,
        p90_latency=p90_latency,
        lanes=lanes,
    )


def estimate_warmup_window_ms(plan: PrefetchPlan) -> int:
    effective_lanes = int(_clamp(plan.lane_count, 1, 8))
    batches = ceil(int(_clamp(plan.budget, 0, 200)) / effective_lanes)
    return batches * int(_clamp(round(plan.p90_latency), 10, 500))
